pkg/sql/sqlstats/persistedsqlstats/compaction_exec.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package persistedsqlstats
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/scheduledjobs"
+	"github.com/cockroachdb/cockroach/pkg/security"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlutil"
+	"github.com/cockroachdb/errors"
+)
+
+// StatsCompactor is responsible for compacting older SQL Stats. It is
+// executed by sql.sqlStatsCompactionResumer.
+type StatsCompactor struct {
+	Settings *cluster.Settings
+
+	initialized bool
+	db          *kv.DB
+	ie          sqlutil.InternalExecutor
+
+	TestingKnobs *StatsCompactorTestingKnobs
+}
+
+// StatsCompactorTestingKnobs can be used to tune the behavior of
+// StatsCompactor in unit tests.
+type StatsCompactorTestingKnobs struct {
+	// DisableFollowerRead disables the use of follower read used in
+	// StatsCompactor. This is used because follower can cause problems in
+	// in unit tests where we immediately start running follower read SQL queries
+	// on system tables that are yet to be created. This is not a scenario that is
+	// possible in real life.
+	DisableFollowerRead bool
+
+	// JobSchedulerEnv overrides the environment to use for scheduled jobs.
+	JobsSchedulerEnv scheduledjobs.JobSchedulerEnv
+}
+
+// Init initializes the StatsCompactor. This method needs to be called before
+// the DeleteOldestEntries method.
+func (c *StatsCompactor) Init(
+	internalEx sqlutil.InternalExecutor, db *kv.DB, disallowFollowerRead bool,
+) {
+	c.db = db
+	c.ie = internalEx
+	c.initialized = true
+	if c.TestingKnobs == nil {
+		c.TestingKnobs = &StatsCompactorTestingKnobs{}
+	}
+	c.TestingKnobs.DisableFollowerRead = disallowFollowerRead
```

Comment:
It's not often that you see non-testing code setting a `TestingKnob` value. Is the idea that your test is going to verify that the boolean made it through to here? If so, use a callback.

Reply:
After reading more, I think it's that you're trying to tell the compactor to disallow follower reads. But, like, that knob is not going to change throughout the life of the cluster so this seems a little funky. Instead just inject the testing knobs in the appropriate places.